### DISCUSSION TITLE
(Unity) How would you program 3d blocks rotating?

### DISCUSSION THREAD

**Discussion by itsjustmonu** [community/community#61097] [Programming Help]
### Body

Hi, I'm trying to learn how to rotate 3d blocks around. I am taking inspiration from the old flash game, Bloxorz. I was wondering how you can code 3d blocks rotating if instead they were something like Tetris blocks. I specifically use Unity but do not know how to even approach this challenge in general.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Hsin-Hung** [comment]
Here are some general ideas on how you can approach 3d blocks rotations:

1. Attach a script to the 3d blocks that you want to rotate.
2. In the script, understand how to use and access [Quaternion](https://docs.unity3d.com/ScriptReference/Quaternion.html) and [Transform.Rotate](https://docs.unity3d.com/ScriptReference/Transform-rotation.html) to manipulate object positions and rotations. Focus on the Transform component. 

Since you mentioned Tetris blocks, you probably want to rotate the block in fixed increments of 90 degrees, so it will be
something like this:
```cs
Vector3 rotationToAdd = new Vector3(0, 90, 0); 
transform.Rotate(rotationToAdd);  // adds 90 degrees to the Y-Axis
```

**Comment by SAMIULLAH-KALHORO-DEV** [comment]
Creating a 3D rotating Tetris-like block in Unity is a fun challenge and a great way to learn about 3D transformations and game development. Here's a general approach to implement this in Unity:

1. Create the Tetris Block Prefab:
   - Start by designing a single Tetris block using Unity's 3D modeling tools or external software.
   - Import the model into Unity and create a Prefab out of it. This will allow you to easily instantiate and manipulate multiple blocks later.

2. Block Movement:
   - Create a script that handles the movement of the block. You can use Unity's built-in physics or write custom movement logic.
   - The block should be able to move left, right, and down, similar to how it moves in the original Tetris game.

3. Block Rotation:
   - Implement a script that handles the rotation of the block. You can do this by applying appropriate rotation transformations to the block's transform.
   - To rotate the block around its center, you can use the `Transform.Rotate` method in Unity.

4. Collision Detection:
   - Ensure that the block doesn't rotate into an invalid position or collide with other blocks or the walls of the play area.
   - Use Unity's collision detection system or write your own custom collision detection logic to prevent illegal rotations.

5. Game Mechanics:
   - Add game mechanics to handle the spawning of new blocks and their interactions with the game environment.
   - When a row is completed, remove the filled row and shift the blocks above it down, just like in the classic Tetris game.

6. Input Handling:
   - Implement input handling to allow players to control the block's movement and rotation using keyboard inputs or touch controls (for mobile).

7. Game Loop:
   - Create a game loop that updates the game state and renders the scene each frame.
   - In the game loop, you'll update the block's position and rotation based on player input and game rules.

8. UI and Graphics:
   - Create a simple UI to display the score and game information.
   - Add some graphics and effects to make the game visually appealing.

Remember that Unity provides numerous resources and tutorials for beginners, including its official documentation and the Unity Learn platform. Be sure to make use of these resources to deepen your understanding of Unity game development.

Start by breaking down the problem into smaller tasks and work on implementing one feature at a time. As you progress, you'll gain a better understanding of game development principles and Unity's capabilities. Good luck, and have fun creating your rotating Tetris-like blocks!
